src/lib/lib.jl

```diff
@@ -210,6 +210,42 @@ end
 @generated pair(::Val{k}, v, _=nothing) where k = :($k = v,)
 @generated pair(::Val{k}, v, ::NamedTuple{keys}) where {k,keys} = k isa Int ? :($(getfield(keys, k)) = v,) : :($k = v,)
 
+# ugly hack to make differentiating `getproperty` infer a lot better
+@generated function _pullback(cx::AContext, ::typeof(literal_getproperty), x, ::Val{f}) where f
+    sig(x) = Tuple{x, typeof(f)}
+    pb_sig(x) = Tuple{cx, typeof(getproperty), x, typeof(f)}
+
+    # either `getproperty` has a custom implementation or `_pullback(cx, ::getproperty, x, f)`
+    # is overloaded directly
+    is_getfield_fallback = which(_pullback, pb_sig(x)) == which(_pullback, pb_sig(Any)) && 
+        which(getproperty, sig(x)) == which(getproperty, sig(Any))
+
+    if is_getfield_fallback
```

Comment:
Another high-level question, what's the reason that this entire case can't simply be something like
```julia
# we've ensured that we're fine to short-circuit and call `getfield` directly,
return :(Zygote._pullback(cx, literal_getfield, x, f))
```
rather than pulling up the IR for that method and pasting it here?

Reply:
The main reason is that we can't (at least that I know of) manually attach backedges to regular expressions, so we need to return a CodeInfo object.  We could return a CodeInfo that just contains that call, but if we already go to lengths like this, we might as well just directly inline the whole thing.

Reply:
This also makes sense to me.